shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/Serialisation/Tripping/CBOR.hs

```diff
@@ -0,0 +1,165 @@
+
+-- | Roundtrip property for values annotated with their serialized form
+--
+-- NOTE: Suppose @a@ consists of a pair of the unannotated value @a'@ and some
+-- 'Lazy.ByteString'. The roundtrip property will fail if that
+-- 'Lazy.ByteString' encoding is not equal to @enc a'@. One way in which this
+-- might happen is if the annotation is not canonical CBOR, but @enc@ does
+-- produce canonical CBOR.
+roundtrip' ::
+  (Eq a, Show a) =>
+  -- | @enc@
+  (a -> Encoding) ->
+  (forall s. Decoder s (Lazy.ByteString -> a)) ->
+  a ->
+  Property
+roundtrip' enc dec a = case deserialiseFromBytes dec bs of
+  Right (bs', a')
+    | Lazy.null bs' ->
+      a === a' bs
+    | otherwise ->
+      counterexample ("left-over bytes: " <> show bs') False
+  Left e ->
+    counterexample (show e) False
+  where
+    bs = toLazyByteString (enc a)
+
+{-------------------------------------------------------------------------------
+  Serialization Properties
+-------------------------------------------------------------------------------}
+
+prop_roundtrip_Addr :: Mock.Addr Monomorphic.ShortHash -> Property
+prop_roundtrip_Addr = roundtrip toCBOR fromCBOR
+
+prop_roundtrip_RewardAcnt :: Mock.RewardAcnt Monomorphic.ShortHash -> Property
+prop_roundtrip_RewardAcnt = roundtrip toCBOR fromCBOR
+
+prop_roundtrip_Block :: Mock.Block Monomorphic.ShortHash -> Property
+prop_roundtrip_Block = roundtrip' toCBOR ((. Full) . runAnnotator <$> fromCBOR)
+
+prop_roundtrip_Header :: Mock.BHeader Monomorphic.ShortHash -> Property
+prop_roundtrip_Header = roundtrip' toCBOR ((. Full) . runAnnotator <$> fromCBOR)
+
+prop_roundtrip_BlockHeaderHash :: Mock.HashHeader Monomorphic.ShortHash -> Property
+prop_roundtrip_BlockHeaderHash = roundtrip toCBOR fromCBOR
+
+prop_roundtrip_Tx :: Mock.Tx Monomorphic.ShortHash -> Property
+prop_roundtrip_Tx = roundtrip' toCBOR ((. Full) . runAnnotator <$> fromCBOR)
+
+prop_roundtrip_TxId :: Mock.TxId Monomorphic.ShortHash -> Property
+prop_roundtrip_TxId = roundtrip toCBOR fromCBOR
+
+prop_roundtrip_TxOut :: Mock.TxOut Monomorphic.ShortHash -> Property
+prop_roundtrip_TxOut = roundtrip toCBOR fromCBOR
+
+prop_roundtrip_BootstrapWitness ::
+  Mock.BootstrapWitness Monomorphic.ShortHash -> Property
+prop_roundtrip_BootstrapWitness = roundtrip' toCBOR ((. Full) . runAnnotator <$> fromCBOR)
+
+prop_roundtrip_LEDGER_PredicateFails :: [STS.PredicateFailure (Mock.LEDGERS Monomorphic.ShortHash)] -> Property
+prop_roundtrip_LEDGER_PredicateFails = roundtrip toCBOR fromCBOR
+
+prop_roundtrip_PrtclState :: STS.PrtclState (Mock.ConcreteCrypto Monomorphic.ShortHash) -> Property
+prop_roundtrip_PrtclState = roundtrip toCBOR fromCBOR
+
+prop_roundtrip_LedgerState :: Mock.LedgerState Monomorphic.ShortHash -> Property
+prop_roundtrip_LedgerState = roundtrip toCBOR fromCBOR
+
+prop_roundtrip_NewEpochState :: Mock.NewEpochState Monomorphic.ShortHash -> Property
+prop_roundtrip_NewEpochState = roundtrip toCBOR fromCBOR
+
+-- TODO
+
+-- roundTripIpv4 :: Property
```

Comment:
are these commented out since they use hedgehog and everything else here is in quickcheck?

Reply:
Yes, sorry, will bring them back once I get back from holiday!